Tasks/AzureResourceGroupDeployment/models/DeployAzureRG.ts

```diff
@@ -106,14 +76,72 @@ export class AzureRGTaskParameters {
         }
     }
 
-    private getARMCredentials(connectedService: string): msRestAzure.ApplicationTokenCredentials {
+    private async getARMCredentials(connectedService: string): Promise<msRestAzure.ApplicationTokenCredentials> {
         var servicePrincipalId: string = tl.getEndpointAuthorizationParameter(connectedService, "serviceprincipalid", false);
         var servicePrincipalKey: string = tl.getEndpointAuthorizationParameter(connectedService, "serviceprincipalkey", false);
         var tenantId: string = tl.getEndpointAuthorizationParameter(connectedService, "tenantid", false);
         var armUrl: string = tl.getEndpointUrl(connectedService, true);
         var envAuthorityUrl: string = tl.getEndpointDataParameter(connectedService, 'environmentAuthorityUrl', true);
-        envAuthorityUrl = (envAuthorityUrl != null) ? envAuthorityUrl : "https://login.windows.net/";
-        var credentials = new msRestAzure.ApplicationTokenCredentials(servicePrincipalId, tenantId, servicePrincipalKey, armUrl, envAuthorityUrl);
+        var environment: string = tl.getEndpointDataParameter(connectedService, 'environment', true);
+        var activeDirectoryResourceId: string = tl.getEndpointDataParameter(connectedService, 'activeDirectoryServiceEndpointResourceId', true);
+        
+        if(environment != null && environment.toLowerCase() == azureStackEnvironment.toLowerCase()) {
+            if(!envAuthorityUrl || !activeDirectoryResourceId) {
+                var endPoint =  await azureStackUtility.initializeAzureStackData({"url":armUrl});
+                envAuthorityUrl = endPoint["environmentAuthorityUrl"];
+                activeDirectoryResourceId = endPoint["activeDirectoryServiceEndpointResourceId"];
+                
+                if(envAuthorityUrl == null) {
+                    throw tl.loc("UnableToFetchAuthorityURL");
+                }
+
+                if(activeDirectoryResourceId == null) {
+                    throw tl.loc("UnableToFetchActiveDirectory");
+                }
+            } 
+        } else {
+            envAuthorityUrl = (envAuthorityUrl != null) ? envAuthorityUrl : "https://login.windows.net/";
+            activeDirectoryResourceId = armUrl;
+        }
+
+        var credentials = new msRestAzure.ApplicationTokenCredentials(servicePrincipalId, tenantId, servicePrincipalKey, armUrl, envAuthorityUrl, activeDirectoryResourceId);
         return credentials;
     }
+
+    public getAzureRGTaskParameters() : Promise<AzureRGTaskParameters> 
+    {
+        return new Promise<AzureRGTaskParameters>(async (resolve, reject) => {
```

Comment:
P2: Consider making getAzureRGTaskParameters async and removing this new Promise